prometheus: Fix metric resets

b-ranto opened the pull request:
This patch changes the way we reset metrics when collecting data in the
prometheus exporter module. With this patch, the exported metrics at any
point in time should align with a freshly started ceph-mgr module.

The patch removes the Metric(s).set methods since these can cause issues
when label values change which can happen e.g. when you remove and
re-deploy an OSD with a different hdd -- the old as well as the new OSD
(with the same ID) would keep showing up in ceph_disk_occupation because
we used Metrics.set to update it instead of Metrics.append.

The patch also changes the reset method name to update to make it more
clear that things have changed in this direction.

Another fix in the patch will clear all the current metrics before
looking at the pending metrics that should be updated. We can do this
since we no longer use the Metric(s).set methods and all the metrics
show up as pending when we are collecting the data. As a result, we can
only call the update method once per data collection.

Signed-off-by: Boris Ranto <branto@redhat.com>

jcsp commented:
Why not get the desired OSD behaviour by using `append` for OSD stats, and keep `set` for the things that really don't go away (like the cluster health)?

b-ranto commented:
@jcsp That would interfere with the other part of the fix where I am clearing all the metrics before updating them in the update method (it would clear the metrics that were modified with the `set` method). The intention there is not to export metrics that no longer exist/could not be collected but still exist in the internal database. This is what should get the exported data in sync with the freshly started ceph-mgr daemon.

When you are collecting the data with prometheus server there is a built-in time-out for the latest metric so even if the latest metric could not be exported you would still see it in the prometheus server database for a while (iirc, the time-out is 5 minutes by default).

jcsp commented:
It seems like at the point all the values are being blown away on each scrape (the new `clear` call on everything in self.metrics), keeping things in "pending" is redundant?  You could just clear all Metric instances at the start of the scrape, and then update Metric.value in place during the scrape.

But I'm not sure how that fits into the original idea of this code, I'm not sure I understand why we had the whole append/set distinction to begin with.

@jan--f any thoughts?

b-ranto commented:
IIRC, we originally only had the `set` method but that changed with a patch that tried to remove old metrics on refresh. That is when the append method was introduced. Eventually, we ended up with a mix of both since the patch did not convert all the metrics.

I suppose we could clear all the metrics at the beginning of the scrape and then `set` them directly during the scrape instead of doing the whole pending mechanism.

jan--f commented:
Originally we had only ```set``` which led to metrics being duplicated when labels changed. To remedy this the ```update``` mechanism was implemented. As @jcsp mentioned, the ```set``` functionality was left for metrics that don't change (which didn't turn out to be as many as I thought).
If we abandon the ```set/update``` distinction (and I'm fine with this), we should also get rid of the pending mechanism, that would just be extra work the module does for no reason. If that is deemed too much effort (see below) changing the offending metrics from ```set``` to ```update/reset``` would also do the trick.
However, going that route (abandon set/update distinction), I'd argue for collecting the metrics as per-request state. Right now every scrape alters the module-global state, i.e. two well timed requests probably interact with each other (metrics being removed by the second request before being formatted and returned by the first). Since we collect all metrics on every request anyway, there is no real reason to keep that state around is there?
I realise this is a bigger effort and not quite the topic of this PR, but I thought its worth bringing it up here.

jcsp commented:
> Since we collect all metrics on every request anyway, there is no real reason to keep that state around is there?

It's possible that many users will do a multi-prometheus setup where each prom server is hitting the Ceph endpoint separately, maybe it makes to do a very short term cache of the generated prometheus result, that times out after ~5 seconds?

jan--f commented:
I was going to propose doing the data update independently from scrapes (using notify), but I feel like that was considered before and ruled out.
A short lived cache sounds like a great solution for this.

b-ranto commented:
The prometheus documentation does not recommend doing any sort of caching unless it takes a long time to scrape a certain metric (they consider 'over a minute' a long time) [1].

The whole scraping for us generally takes less than a second and I am not sure why anyone would want to run multiple Prometheus servers in their setup (outside of a test environment). From what I understand, the aim is towards a centralized Prometheus server for all of your data.

We may want to introduce a certain locking methodology on scrapes though since these update the global_instance and can in fact overlap, especially when you are doing them manually (with web browser).

I am in favour of getting rid of the pending mechanism since that does not seem to make much sense anymore.

[1] https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling

b-ranto commented:
FYI: I have re-worked the commit to remove the whole `pending` mechanism, introduce caching and added locking to avoid overwriting the data structures. I have also cleaned the schema OrderedDict that is no longer used in the code.

jan--f commented on src/pybind/mgr/prometheus/module.py:
```diff
@@ -110,40 +110,30 @@ def health_status_to_number(status):
+        self.metrics[key].set(value, labels)
 
-    def reset(self):
+    def all(self):
```
This function call could be avoidedby simply returning the ```metrics``` object.

b-ranto replied:
I was thinking of removing the Metrics class altogether since now it is basically a simple dictionary. What about doing that?

b-ranto commented on src/pybind/mgr/prometheus/module.py:
```diff
@@ -110,40 +110,30 @@ def health_status_to_number(status):
-        self.pending = {}
 
-    def set(self, key, value, labels=('',)):
+    def clear(self):
```
That would mean that we would have to setup static metrics (and all the other metrics) on each scrape. I did it this way to avoid the overhead of creating all the Metric instances on each scrape.

b-ranto commented on src/pybind/mgr/prometheus/module.py:
```diff
@@ -688,14 +680,31 @@ def index(self):
             def metrics(self):
-                if global_instance().have_mon_connection():
-                    metrics = global_instance().collect()
+                inst = global_instance()
```
I thought `instance` was a keyword or a module name for some reason (I guess I mixed it up in my mind with the `isinstance` method or the `object` class). It does not seem to be the case so I can change it to `instance`.

b-ranto commented:
@jan-f The `metrics` and `collect_cache` are slightly different. The former contains the classes while the latter contains the formatted output of the metrics. I keep the `collect_cache` to avoid re-formatting the `metrics` on each request while I keep the original `metrics` to avoid re-constructing all the `Metric` instances. We could `clear` them after they are formatted as a `collect_cache` though.

b-ranto commented:
I have pushed two more commits. The first one changes the `inst` variable to `instance`. The second one optimizes output formatting for prometheus (it is faster to use `join` then the standard `+` operator for multiple values).

b-ranto commented:
I have pushed two more commits for further optimizations. The first removes the `Metrics` class to avoid creation of unnecessary `Metric` classes and the second one moves the formatting to the `collect` method so that we could clear the memory.

jan--f commented on src/pybind/mgr/prometheus/module.py:
```diff
@@ -551,35 +525,40 @@ def collect(self):
-                        ("ceph_daemon",),
-                    ))
-                    self.metrics.append(_path, value, (daemon,))
+                    if _path not in self.metrics:
```
Minor nit: The ```Total``` and ```Count``` metrics could be setup in the same if block. I assume they always come in this pair? That would save one conditional in line 538.

b-ranto replied:
It would not really save the conditional, we would have to use `_path1 not in self.metrics or _path2 not in self.metrics` to be sure that the conditional is working properly (*). Fortunately, both of these are simple hash table lookups so it is not too bad performance-wise.

(*) I just have a bad feeling that someone somewhere will change a `<stg>_count` or `<stg>_sum` perf counter into a histogram that is represented by the long running avgs as a pair of `<stg>_sum` and `<stg>_count` and the code would break. It would also be pretty difficult to debug if it really happened.

b-ranto commented:
I have pushed two more commits for the issues that I noticed while testing this -- we have never set the `collect_time` so the caching would never apply and the response header for the cached data was not being set.

b-ranto commented:
jenkins retest this please

b-ranto commented:
all green: http://pulpito.ceph.com/branto-2018-07-17_20:20:54-rados:mgr-wip-metric-reset-distro-basic-smithi/

b-ranto commented:
jenkins retest this please
